ably/realtime_conn_test.go

```diff
@@ -139,3 +140,58 @@ func TestRealtimeConn_AuthError(t *testing.T) {
 		t.Fatal("Close(): want err != nil")
 	}
 }
+
+func TestRealtimeConn_ReceiveTimeout(t *testing.T) {
+	t.Parallel()
+
+	const maxIdleInterval = 20
+	const realtimeRequestTimeout = 10 * time.Millisecond
+
+	in := make(chan *proto.ProtocolMessage, 16)
+	out := make(chan *proto.ProtocolMessage, 16)
+
+	connected := &proto.ProtocolMessage{
+		Action:       proto.ActionConnected,
+		ConnectionID: "connection-id",
+		ConnectionDetails: &proto.ConnectionDetails{
+			MaxIdleInterval: maxIdleInterval,
+		},
+	}
+	in <- connected
+
+	app, client := ablytest.NewRealtimeClient(&ably.ClientOptions{
+		Dial:                   ablytest.MessagePipe(in, out),
+		RealtimeRequestTimeout: 10 * time.Millisecond,
+		NoConnect:              true,
+	})
+	defer safeclose(t, app)
+
+	states := make(chan ably.State, 10)
+	client.Connection.On(states, ably.StateConnConnected, ably.StateConnFailed)
+
+	client.Connection.Connect()
+
+	var state ably.State
+
+	select {
+	case state = <-states:
+	case <-time.After(10 * time.Millisecond):
+		t.Fatal("didn't receive state change event")
+	}
+
+	if expected, got := ably.StateConnConnected, state.State; expected != got {
+		t.Fatalf("expected %v, got %v", expected, got)
+	}
+
+	leeway := 10 * time.Millisecond
+	select {
+	case state = <-states:
+	case <-time.After(realtimeRequestTimeout + time.Duration(maxIdleInterval)*time.Millisecond + leeway):
+		t.Fatal("didn't receive state change event")
+	}
+
+	// TODO: Should be Disconnected, not Failed
```

Comment:
Is there an issue for this? Can we link it here?

Reply:
```suggestion
	// TODO: Should be Disconnected, not Failed
	// Part of https://ably.atlassian.net/browse/FEA-391
```